doc/LAYERS.org

```diff
@@ -0,0 +1,588 @@
+MELPA recipes (through the Quelpa package). See the [[https://github.com/milkypostman/melpa#recipe-format][MELPA documentation]] for more
+information about recipes.
+
+Packages may be /excluded/ by setting the =:excluded= property to true. This
+will prevent the package from being installed even if it is used by another
+layer.
+
+For each included package, you may define one or more of the following
+functions, which are called in order by Spacemacs to initialize the package.
+
+1. =<layer>/pre-init-<package>=
+2. =<layer>/init-<package>=
+3. =<layer>/post-init-<package>=
+
+It is the responsibility of these functions to load and configure the package in
+question. Spacemacs will do nothing other than download the package and place it
+in the load path for you.
+
+*Note:* A package will not be installed unless at least one layer defines an
+=init= function for it. That is to say, in a certain sense, the =init= function
+does mandatory setup while the =pre-init= and =post-init= functions do optional
+setup. This can be used for managing cross-layer dependencies, which we will
+discuss later.
+
+** config.el, funcs.el and keybindings.el
+These files contain no magic variables or functions, merely code that is
+executed. The intended use cases are
+
+- config.el :: configuration code that is independent of packages, declaration
+     layer variables, and so on
+- funcs.el :: definition of general utility functions
+- keybindings.el :: general keybindings
+
+The word /general/ here means /independent of any package/. Since the end user
+can exclude an arbitrary set of packages, you cannot be sure that, just because
+your layer includes a package, that package will necessarily be loaded. For this
+reason, code in these files must be generally safe, regardless of which packages
+are installed.
+
+More on this in the next section.
+
+* The Spacemacs loading process
+The Spacemacs loading process can be summarized as follows:
+
+1. A handful packages are installed and loaded as part of the bootstrap process.
+   Among these packages, of particular interest to us, is =use-package=. All
+   subsequent code in the loading process may make use of it.
+2. Spacemacs goes through all the enabled layers and evaluates their files. The
+   changes introduced by =config.el= are thus applied, but nothing happens from
+   =packages.el=, since these files only define functions and variables.
+3. Spacemacs checks which packages should be downloaded and installed. To be
+   installed, a package must be
+   - included by a layer that the user has enabled,
+   - not be excluded by any other layer that the user has enabled,
+   - not be excluded by the user herself, and
+   - there must be at least one =<layer>/init-<package>= function defined for
+     it.
+   Alternatively, if a package is part of the end user's
+   =dotspacemacs-additional-packages=, it will also be installed.
+4. All packages which should be installed are, and all packages which are
```

Comment:
I meant *All packages which should be installed are installed*, but I guess I can write that explicitly.